src/hmac.rs

```diff
@@ -245,11 +268,91 @@ impl Key {
+
+    /// Verifies that the HACM of `data` with key equal `tag`.
+    ///
+    /// Use [`verify_parts`] if the data is not in a contiguous memory region.
+    ///
+    /// Conceptually, this will calculates the HMAC of `data` using the signing
+    /// key, and verifies whether the resultant value equals `tag`, in one step
+    /// and without making the resulting tag available. The comparison will be
+    /// done in constant time to prevent timing attacks.
+    ///
+    /// [`verify_parts`]: #method.verify_parts
+    pub fn verify(&self, data: &[u8], tag: &[u8]) -> Result<(), error::Unspecified> {
+        self.sign(data).check(tag)
+    }
+
+    /// Sign data distributed in multiple memory regions.
+    ///
+    /// This is effectively the same as if signing the concatenation of all byte
+    /// slices returned by the iterator. Use [`Context`] to calculate HMACs
+    /// where the input is not available at once.
+    ///
+    /// It is generally not safe to implement HMAC verification by comparing the
+    /// return value of `sign` to a tag. Use [`verify_parts`] for verification
+    /// instead.
+    ///
+    /// [`verify_parts`]: #method.verify_parts
+    /// [`Context`]: ./struct.Context.html
+    pub fn sign_parts<'a, I>(&self, parts: I) -> Tag
+        where I: IntoIterator<Item=&'a [u8]>,
+    {
+        let mut ctx = Context::with_key(self);
+        parts.into_iter().for_each(|part| ctx.update(part));
+        ctx.sign()
+    }
+
+    /// Verify data distributed in multiple memory regions.
+    ///
+    /// This is effectively the same as if verifying the concatenation of all
+    /// byte slices returned by the iterator.
+    ///
+    /// It is not possible to create verification context that does not verify
+    /// the tag in one step after gather all inputs. This avoid accidentally
+    /// processing parts of the input while under the *wrong* impression that
+    /// they were already verified.
+    ///
+    /// Conceptually, this will calculates the HMAC of `data` using the signing
+    /// key, and verifies whether the resultant value equals `tag`, in one step
+    /// and without making the resulting tag available. The comparison will be
+    /// done in constant time to prevent timing attacks.
+    pub fn verify_parts<'a, I, T>(&self, parts: I, tag: &T) -> Result<(), error::Unspecified>
+    where
+        I: IntoIterator<Item=&'a [u8]>,
+        Tag: PartialEq<T>,
+        T: ?Sized,
+    {
+        self.sign_parts(parts).check(tag)
+    }
 }
 
 /// A context for multi-step (Init-Update-Finish) HMAC signing.
 ///
-/// Use `sign` for single-step HMAC signing.
+/// Use [`Key::sign`] for single-step HMAC signing and [`Key::sign_parts`] for signle step signing
+/// of data that is not in a contiguous memory region. Note that there is no verification parallel
+/// to [`Context`] to avoid misusing data that is not already verified.
+// Note: No verify. This is to require all data to be verified and discourage processing parts of
```

Comment:
That note is to the implementors only (no doc-comment(. During the process of the issue there was some confusion which methods should be create and this is justification why some are not part of this PR and should never be.

Reply:
I was just saying that I thought the doc comment was clear enough to a reader/reviewer of the code as well. So I didn't think the extra comment added anything (it actually confused me initially).

Reply:
Ah, good to know. Do you think there's a better way to refer to the commit or to further justification? This could of course also be relevant in the general documentation and there is no inherent reason to split the information.

Reply:
If it's needed for external users of the API, it should be explained in a doc comment (which you already do). If it's only relevant to implementers/code-review, the context should be in the PR, so that the `git blame` will point to the right stuff.

Reply:
The problem with `git blame` is that it points to a commit for a line in the file. But the comment is about lines *that should not exist* in the module in the first place. It should also be some place where it would catch the attention of a programmer trying to add them and reviewers who'd inspect changes of that manner, for both of these scenarios usage of `git blame` doesn't seem likely enough.

I suppose the public API doc would be enough, as it would be checked for consistency with the implementation.